os/bluestore: expand lttng tracepoints, improve fio_ceph_objectstore backend

athanatos opened the pull request:
Adds bluestore tracepoints for recording a variety of initial conditions for sampled ios including:
- rocksdb properties like estimate-pending-compaction-bytes, num-running-compactions etc
- current throttle values at IO initialization
- duration of each bluestore stage

Also updates the fio_ceph_objectstore backend to optionally emit json formatted perf counters to a file as well as support for running multiple jobs and varying the throttle limits during a run.

The intention is that this functionality as well https://github.com/ceph/cbt/pull/190 will enable somewhat quicker iteration on modifications to bluestore intended to improve throttle behavior or IO latency variability.

athanatos commented:
retest this please.

aclamk commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -482,6 +614,45 @@ Job::~Job()
+      int r = engine->cct->_conf.set_val(
+	"bluestore_throttle_bytes",
+	val,
+	nullptr);
```
How does value "bluestore_throttle_bytes" propagate from client (FIO) to server (osd/bluestore)?

athanatos replied:
In this instance, the bluestore instance is in the same instance and this backend interacts with it directly.  There's no osd or objecter.  Modifying cct there directly updates bluestore.

athanatos replied:
This doesn't seem problematic to me.  bluestore_throttle_bytes and _deferred_bytes are the same throttles the osd uses to regulate how much work bluestore will accept before blocking (indeed, the purpose here is essentially to determine what those throttles should be set to).  As such, there's already support for modifying them on the fly via injectargs etc.  Adding an additional interface seems superfluous.

athanatos replied:
Also, there's no admin socket here since there's no osd.  The admin socket is created by the osd and simply uses this config modification mechanism.

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -14034,6 +14075,168 @@ void BlueStore::log_latency_fn(
+  mono_clock::time_point start_throttle_acquire)
+{
+  pending_kv_ios += txc.ios;
+  if (txc.deferred_txn) {
```
```
  <kind>InvalidRead</kind>
  <what>Invalid read of size 8</what>
  <stack>
    <frame>
      <ip>0xAEBB5D</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::BlueStoreThrottle::complete(BlueStore::TransContext&amp;)</fn>
      <dir>/usr/src/debug/ceph-15.0.0-4218-gcff0a5a/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>14245</line>
    </frame>
    <frame>
      <ip>0xB3C322</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::_txc_finish(BlueStore::TransContext*)</fn>
      <dir>/usr/src/debug/ceph-15.0.0-4218-gcff0a5a/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>10623</line>
    </frame>
    <frame>
      <ip>0xB3DD0B</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::_txc_state_proc(BlueStore::TransContext*)</fn>
      <dir>/usr/src/debug/ceph-15.0.0-4218-gcff0a5a/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>10327</line>
    </frame>
    <frame>
      <ip>0xB3F334</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::_kv_finalize_thread()</fn>
      <dir>/usr/src/debug/ceph-15.0.0-4218-gcff0a5a/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>11153</line>
    </frame>
    <frame>
      <ip>0xB9927C</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::KVFinalizeThread::entry()</fn>
      <dir>/usr/src/debug/ceph-15.0.0-4218-gcff0a5a/src/os/bluestore</dir>
      <file>BlueStore.h</file>
      <line>1852</line>
    </frame>
```

see http://pulpito.ceph.com/kchai-2019-08-25_05:28:04-rados-wip-kefu-testing-2019-08-23-1454-distro-basic-smithi/4251294/

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-08-30_13:08:37-rados-wip-kefu2-testing-2019-08-29-0935-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-08-29_09:35:06-rados-wip-kefu2-testing-2019-08-29-0935-distro-basic-mira/

tchaikov commented:
@athanatos needs rebase.

athanatos commented:
@tchaikov rebased

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-09-07_14:44:52-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-08_13:41:17-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-smithi/

athanatos commented:
retest this please

athanatos commented:
@aclamk Had a chance to review this?

athanatos commented:
@aclamk Or should I ask someone else for a review?

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -43,6 +43,16 @@
+#if defined(WITH_LTTNG)
+#define TRACEPOINT_DEFINE
+#define TRACEPOINT_PROBE_DYNAMIC_LINKAGE
+#include "tracing/bluestore.h"	// 
```
coud remove the trailing `//` ?

tchaikov commented on src/common/ceph_time.h:
```diff
@@ -503,6 +503,20 @@ struct converts_to_timespec<Clock, std::void_t<decltype(
 constexpr bool converts_to_timespec_v = converts_to_timespec<Clock>::value;
 
+template<typename Target, typename T>
+static double to_seconds(T t) {
```
might want to s/Target/Rep/ and  s/double/Rep/ ?

tchaikov commented on src/common/ceph_time.h:
```diff
@@ -503,6 +503,20 @@ struct converts_to_timespec<Clock, std::void_t<decltype(
+}
+
+template<typename Target, typename T>
+static uint64_t to_microseconds(T t) {
```
might want to s/Target/Rep/ and s/uint64_t/Rep/ ?

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -14784,8 +14925,44 @@ void BlueStore::BlueStoreThrottle::finish_start_transaction(
+#if defined(WITH_LTTNG)
+void BlueStore::BlueStoreThrottle::complete_kv(TransContext &txc)
+{
+  pending_kv_ios -= 1;
```
nit, could put `--pending_kv_ios;`

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -14784,8 +14925,44 @@ void BlueStore::BlueStoreThrottle::finish_start_transaction(
+void BlueStore::BlueStoreThrottle::complete(TransContext &txc)
+{
+  if (txc.deferred_txn) {
+    pending_deferred_ios -= 1;
```
ditto

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -14784,8 +14925,44 @@ void BlueStore::BlueStoreThrottle::finish_start_transaction(
+  }
+  if (txc.tracing) {
+    utime_t now = ceph_clock_now();
+    double usecs = (now.to_nsec()-txc.start.to_nsec())/1000;
```
shall we cast `now.to_nsec()-txc.start.to_nsec()` to double before dividing it by 1000 ?

also, might want to add spaces around the operators.

tchaikov commented on src/os/bluestore/BlueStore.h:
```diff
@@ -28,12 +30,14 @@
 #include <boost/intrusive/set.hpp>
 #include <boost/functional/hash.hpp>
 #include <boost/dynamic_bitset.hpp>
+#include <boost/circular_buffer.hpp>
```
ditto.

tchaikov commented on src/os/bluestore/BlueStore.h:
```diff
@@ -20,6 +20,8 @@
 
 #include <atomic>
+#include <chrono>
+#include <ratio>
```
nit, please order the `#include`s alphabetically.

tchaikov commented on src/os/bluestore/BlueStore.h:
```diff
@@ -28,12 +30,14 @@
 #include "include/ceph_assert.h"
 #include "include/unordered_map.h"
 #include "include/mempool.h"
+#include "include/hash.h"
```
ditto.

tchaikov commented on src/tracing/bluestore.tp:
```diff
@@ -0,0 +1,148 @@
+    TP_ARGS(
+        uint32_t, sequencer_id,
+        uint64_t, tid,
+	uint8_t, sync,
```
might want to use spaces as well?

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -495,6 +505,29 @@ Job::Job(Engine* engine, const thread_data* td)
+    }
+  }
+  if (o->check_files) {
+    derr << "fio_ceph_objectstore checked " << checked << " files"<< dendl;
```
i think we need to use `dout(0)` for non-error output.

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -495,6 +505,29 @@ Job::Job(Engine* engine, const thread_data* td)
         throw std::system_error(r, std::system_category(), "job init");
       }
     }
+    if (o->check_files) {
```
shall we just `continue` if `o->preallocate_files` is false?

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -495,6 +505,29 @@ Job::Job(Engine* engine, const thread_data* td)
+      }
+    }
+    if (o->check_files || o->preallocate_files) {
+      ++checked;
```
i don't follow here. if we stat an object only if `check_files` is true, why shall we increment `checked` if `o->preallocate_files` ?

athanatos replied:
I reused the variable, I'll rename it.

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -441,22 +469,58 @@ struct Object {
+  mono_clock::time_point last = ceph::mono_clock::zero();
+  unsigned index = 0;
+
+  static vector<unsigned> parse_throttle_str(const char *p) {
```
nit, could use `ceph::for_each_substr()`.

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -441,22 +469,58 @@ struct Object {
+    std::cerr << "throttle: " << in << std::endl;
+
+    size_t pos = 0;
+    while (pos != std::string::npos && pos < in.size()) {
```
s/std::string::npos/in.npos/

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -550,6 +614,45 @@ Job::~Job()
+    }
+    if (!deferred_throttle_values.empty()) {
+      std::string val = std::to_string(deferred_throttle_values[index % dtvals]);
+      std::cerr << "Setting bluestore_deferred_throttle_bytes to " << val << std::endl;
```
s/cerr/cout/

athanatos replied:
fio outputs stuff on cout, this is easier.

tchaikov commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -550,6 +614,45 @@ Job::~Job()
+    unsigned total = tvals * dtvals;
+    if (!throttle_values.empty()) {
+      std::string val = std::to_string(throttle_values[index % tvals]);
+      std::cerr << "Setting bluestore_throttle_bytes to " << val << std::endl;
```
s/cerr/cout/

athanatos replied:
fio outputs stuff on cout, this is easier.

athanatos commented on src/CMakeLists.txt:
```diff
@@ -332,6 +332,12 @@ set(libcommon_files
 set_source_files_properties(ceph_ver.c
   APPEND PROPERTY OBJECT_DEPENDS ${CMAKE_BINARY_DIR}/src/include/ceph_ver.h)
+
+if(WITH_LTTNG AND WITH_EVENTTRACE)
```
That sounds good, but looking at config-h.in.cmake I'm not clear on how to do it.  Are there instructions you can point me to?

tchaikov replied:
sure. see https://cmake.org/cmake/help/latest/command/configure_file.html . `WITH_LTTNG` is defined using `#cmakedefine` .

athanatos replied:
Actually, I guess the problem is EVENTTRACE requires LTTNG.  Is there a way to simply error if WITH_LTTNG is not enabled but WITH_EVENTTRACE is?

athanatos replied:
Or to simply set WITH_LTTNG if WITH_EVENTTRACE?

athanatos commented on src/CMakeLists.txt:
```diff
@@ -433,7 +433,7 @@ install(
 if(${WITH_LTTNG})
   add_subdirectory(tracing)
-  add_dependencies(common-objs oprequest-tp)
+  add_dependencies(common-objs tracepoint_libraries)
```
I think the problem is that the only target which builds them otherwise is vstart.  It seemed simpler to do it here rather than to add the specific -tp targets to each place.

tchaikov replied:
> only target which builds them otherwise is vstart

yes. but what the problem/pain point this change is trying to address?

athanatos replied:
Ok, I think it was left over from before I went through and added in the EVENTTRACE dependencies everywhere else.  I've removed it now. Sorry!

athanatos commented:
Ok, I think I've addressed your comments.  For the cmake bit, I added a cmakedefine for EVENTTRACE, made it depend on LTTNG, and cleaned up all of the code and cmake conditionals to only check EVENTTRACE.

athanatos commented:
@tchaikov ^

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -332,6 +332,11 @@ set(libcommon_files
   APPEND PROPERTY OBJECT_DEPENDS ${CMAKE_BINARY_DIR}/src/include/ceph_ver.h)
+
+if(WITH_EVENTTRACE)
+  message(STATUS " Using EventTrace class.")
```
might want to drop this feedback as it's not an option/feature enabled/identified by us. it's enabled by user.

tchaikov commented on src/librbd/CMakeLists.txt:
```diff
@@ -142,6 +142,12 @@ set(librbd_internal_srcs
   ${CMAKE_SOURCE_DIR}/src/common/ContextCompletion.cc)
 
+if(WITH_EVENTTRACE)
+  message(STATUS " Using EventTrace class.")
```
this feedback is not necessary.

tchaikov commented on src/msg/async/ProtocolV1.cc:
```diff
@@ -977,12 +974,12 @@ CtPtr ProtocolV1::handle_message_footer(char *buffer, int r) {
     utime_t ltt_processed_stamp = ceph_clock_now();
     double usecs_elapsed =
-        (ltt_processed_stamp.to_nsec() - ltt_recv_stamp.to_nsec()) / 1000;
+        (ltt_processed_stamp.to_nsec() - recv_stamp.to_nsec()) / 1000;
```
cast to double before dividing by 1000.

tchaikov commented:
please note, some of my comments are folded by GitHub web UI.

tchaikov commented:
i still think

- https://github.com/ceph/ceph/pull/29674/files#r327635455
- https://github.com/ceph/ceph/pull/29674#discussion_r328400221
- https://github.com/ceph/ceph/pull/29674#discussion_r327633674
- https://github.com/ceph/ceph/pull/29674#discussion_r327638803

are valid concerns.

athanatos commented:
Sorry, I did indeed miss those.

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -11650,7 +11643,7 @@ void BlueStore::_kv_sync_thread()
 	} else {
 	  ceph_assert(txc->state == TransContext::STATE_KV_SUBMITTED);
-	  txc->log_state_latency(logger, l_bluestore_state_kv_queued_lat);
+	  ceph_assert(0 == "should be configured to be impossible");
```
nit, why bother checking `ceph_assert(txc->state == TransContext::STATE_KV_SUBMITTED);` if all transactions come from `kv_queue_unsubmitted`.

athanatos replied:
Yeah, I messed this up in one of the rebases, I'll fix it.

athanatos replied:
That loop is actually over kv_committing which comes from kv_queue.  My assert there is a mistake, I added it during debugging to confirm that bluestore was doing what I thought it was doing and failed subsequently to remove it.  That path (and the first assert) reflect the bluestore_sync_submit_transaction path in _txc_state_proc by which the transaction will be put into the kv_queue (though not kv_queue_unsubmitted) *after* being synchronously submitted to the kv backend.  kv_queue_unsubmitted appears to be a subset of kv_queue consisting of those kv operations which have not yet been submitted.

athanatos replied:
Actually, looking at it again, l_bluestore_state_kv_queued_lat seems to be intended to track removal from the queue, not application.  I'm going to move that back to here from _txc_apply_kv.

aclamk commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -482,6 +607,45 @@ Job::~Job()
+    engine->cct->_conf.apply_changes(nullptr);
+    index++;
+    index %= tvals * dtvals;
+    last = ceph::mono_clock::now();
```
Unnecessary modulo trim of index %= tvals * dtvals . All other operations are modulo already.

athanatos replied:
The intention is to avoid a wraparound on a long running test.

aclamk commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -281,38 +333,57 @@ struct Engine {
+		    << ", exception: " << e.what()
+		    << std::endl;
+	}
+      }
```
Maybe spaces instead of tabs?

athanatos replied:
Most ceph files are spaced like that, I'm afraid.  It's specified in the vim and emacs strings at the top of the file.

athanatos commented on src/common/options.cc:
```diff
@@ -5526,6 +5526,14 @@ std::vector<Option> get_global_options() {
+    Option("bluestore_throttle_trace_rate", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0)
+    .set_description(""),
+
```
I'm not sure what you mean.  I've added a description though.

athanatos commented:
retest this please

athanatos commented:
retest this please

athanatos commented:
Test run: http://pulpito.ceph.com/sjust-2019-09-29_21:12:10-rados-wip-sjust-testing2-distro-basic-smithi/
Rerun: http://pulpito.ceph.com/sjust-2019-09-30_17:01:52-rados-wip-sjust-testing2-distro-basic-smithi/
Remaining failures appear unrelated.
